### PR TITLE
[925] Test that only valid signals are generated by LEDGER STS

### LINKED ISSUE #925: Test that only valid signals are generated
To start with we can test the signal generation of the LEDGER Trace generator.
 
However, this needs to be done for each new Trace generator yet to be created.

See `onlyValidSignalsAreGenerated` in "cardano-ledger-specs/byron/semantics/executable-spec/src/Control/State/Transition/Generator.hs" and uses of this property in the Byron property tests.

### PR DESCRIPTION
**Pull request by uroboros**
Closes #925

### CODE REVIEW

**Comment by uroboros** [comment]
@dnadales I notice that the `relevantKValues` coverage tests are failing - I can change that as part of this PR - when you review, it would be good to know what could have shifted that coverage downward...

Screenshot 2019-10-22 at 16 12 37

File: shelley/chain-and-ledger/executable-spec/test/Rules/ClassifyTraces.hs [hunk 1]
```diff
@@ -79,3 +71,10 @@ lenRatio :: ([a] -> [b]) -> [a] -> Double
 lenRatio f xs
   = ratioInt (length (f xs))
              (length xs)
+
+onlyValidLedgerSignalsAreGenerated :: Property
+onlyValidLedgerSignalsAreGenerated =
+  withTests 200 $
+    onlyValidSignalsAreGeneratedForTrace traceGen
+  where
+    traceGen = traceOfLengthWithInitState @LEDGER 100 mkGenesisLedgerState
```
Comment (dnadales) [inline shelley/chain-and-ledger/executable-spec/test/Rules/ClassifyTraces.hs]: I should have asked this before, but can't the genesis state be generated from the environment?

**Comment by dnadales** [comment]
> @dnadales I notice that the `relevantKValues` coverage tests are failing - I can change that as part of this PR - when you review, it would be good to know what could have shifted that coverage downward...
> 
> Screenshot 2019-10-22 at 16 12 37

I don't see anything that could have decreased the coverage. Do you get this failure consistently?

File: byron/semantics/executable-spec/src/Control/State/Transition/Generator.hs [hunk 2]
```diff
@@ -602,8 +603,18 @@ onlyValidSignalsAreGenerated
...
-  tr <- forAll (trace @s maximumTraceLength)
+onlyValidSignalsAreGenerated maximumTraceLength =
+  onlyValidSignalsAreGeneratedForTrace (trace @s maximumTraceLength)
+
+-- | Check that the signal generator of 's' only generate valid signals.
+onlyValidSignalsAreGeneratedForTrace
+  :: forall s
+   . (HasTrace s, Show (Environment s), Show (State s), Show (Signal s), HasCallStack)
+  => Gen (Trace s)
+  -- ^ Maximum trace length.
```
Comment (dnadales) [inline byron/semantics/executable-spec/src/Control/State/Transition/Generator.hs]: Uh? Wrong comment?

**Comment by dnadales** [comment]
I've been running

```sh
while stack test cs-ledger:test:ledger-rules-test --ta '-p "Relevant k values"'; do : ; done
```

for a while on your branch, but I cannot reproduce the failure :man_shrugging: Make sure you stay away from electrical storms ...

**Comment by dnadales** [comment]
Argh! Hydra is failing due to `mingw` https://hydra.iohk.io/build/1264250

File: shelley/chain-and-ledger/executable-spec/test/Rules/ClassifyTraces.hs (hunk 1 above)
Comment (uroboros) [reply shelley/chain-and-ledger/executable-spec/test/Rules/ClassifyTraces.hs]: In this case the genesis state (utxo0) is independent of the environment

File: byron/semantics/executable-spec/src/Control/State/Transition/Generator.hs (hunk 2 above)
Comment (uroboros) [reply byron/semantics/executable-spec/src/Control/State/Transition/Generator.hs]: 🙏

**Comment by uroboros** [comment]
> I've been running
> 
> ```shell
> while stack test cs-ledger:test:ledger-rules-test --ta '-p "Relevant k values"'; do : ; done
> ```
> 
> for a while on your branch, but I cannot reproduce the failure 🤷‍♂ Make sure you stay away from electrical storms ...

I shut down the fusion reactor and it runs ok now! ;)

File: byron/semantics/executable-spec/src/Control/State/Transition/Generator.hs [hunk 3]
```diff
@@ -330,7 +332,7 @@ genTraceOfLength aTraceLength profile env st0 aSigGen =
           loop (d - 1) sti acc
         Just sig ->
           case applySTS @s (TRC(env, sti, sig)) of
-            Left _err  -> loop (d - 1) sti acc
+            Left _err  -> D.trace (show _err) (loop (d - 1) sti acc)
```
Comment (mgudemann) [inline byron/semantics/executable-spec/src/Control/State/Transition/Generator.hs]: as above

**Comment by dnadales** [comment]
So yeah, with enough time and patience the tests will fail

> I've been running
> 
> ```shell
> while stack test cs-ledger:test:ledger-rules-test --ta '-p "Relevant k values"'; do : ; done
> ```
> 
> for a while on your branch, but I cannot reproduce the failure  Make sure you stay away from electrical storms ...

After a while it failed, but failing in a different coverage requirement.

image

So since the cover percentages are arbitrary anyway, you might want to make them all 5%. The important thing is that *we are* generating those cases (I guess).

File: byron/semantics/executable-spec/src/Control/State/Transition/Generator.hs [hunk 4]
```diff
@@ -92,6 +93,7 @@ import qualified Hedgehog.Extra.Manual as Manual
 
 import           Test.Goblin (Goblin (..), GoblinData, SeedGoblin (..))
 
+import qualified Debug.Trace as D
```
[the remark this answers is not in the document]
Comment (uroboros) [reply byron/semantics/executable-spec/src/Control/State/Transition/Generator.hs]: yes, thanks! will remove

**Comment by dnadales** [comment]
Did we merge this without hydra passing?

**Pull request merged**